src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+.. :Authors: - Cyrus Omar, Abhitej Ganta
 
 .. title:: Programming Assistants
 
 Overview
 ========
 
-TODO: Overview
+Development software has improved far beyond what was initially conceived. **Programming assistants**, tools that are integrated into this software to enable more effective programming, are leading the charge with new initiatives and improvements every year. Why do we need programming assistants? As more complex and time-consuming projects arise, assistants can significantly reduce the time and effort a programmer exhausts during development.
+
+Programming assistants can be highly sophisticated themselves, with most using advanced artificial intelligence, machine learning and natural language processing models. However, most models can be represented by the following diagram:
+
+-- Insert Diagram Here --
+
+where the input can be range from syntactical structure (abstract syntax trees, parsed text) to program runtime data. The chosen model is done so based on accuracy and runtime comparisons between various hypothesized models. Even after the model is chosen, it is further evaluated and tuned. Once a model is chosen and trained on input data, new data is introduced for prediction and the results are published. The main areas of research for programming assistants are discussed further below.
 
 Refactoring
 ===========
 
+**Refactoring** refers to the process of restructuring existing code without changing the program's behavior and is a large area of interest in programming assistant research. It is intended to improve the nonfunctional attributes of a program including code readability, reduced complexity and source code maintainability. Typically, code refactoring applies small changes to various areas of a program and most development environments have already integrated support for basic refactoring.
+
+Refactoring is motivated by indications of characteristics in a program of a deeper problem. This is known as **code smell**. Detecting problems early and refactoring source code can save developers time and effort debugging in the future. The main benefits of refactoring can be categorized into either **maintainability**: Fixing issues is very easy because the code is easy to read and easy to understand, or **extensibility**: Extending the capabilities of the application is very easy because the codes uses recognizable design patterns and is flexible.
+
+Software engineers at Microsoft are currently developing and integrating a refactoring assistant that recognizes editing patterns in source code and suggest other edits :cite:`miltner2019fly`.
+
+Consider a corpus of code where every line is similar like the following:
+.. code-block:: python
+  var1 = ExampleFunction(param1)
+  var2 = ExampleFunction(param2)
+  var3 = ExampleFunction(param3)
+  var4 = ExampleFunction(param4)
+
+A new function is then written to initialize the created variables so the code must be changed to call the new function:
```

Comment:
This isn't the best example since its just a simple renaming. For describing Blue Pencil, maybe use the example from the video or just link to that blog post.